examples_tests/60.ClusteredRendering/cull_common.glsl

```diff
@@ -0,0 +1,247 @@
+	planes[5].w = dot(planes[5].xyz, p0);
+
+	return planes;
+}
+
+bvec3 and(in bvec3 a, in bvec3 b)
+{
+	return bvec3(a.x && b.x, a.y && b.y, a.z && b.z);
+}
+
+bvec3 or(in bvec3 a, in bvec3 b)
+{
+	return bvec3(a.x || b.x, a.y || b.y, a.z || b.z);
+}
+
+float projectedSphericalVertex(in vec3 origin, in vec3 planeNormal, in vec3 pos)
+{
+	return dot(normalize(pos - origin), planeNormal);
+}
+
+vec3 findFarthestPointOnConeInDirection(in vec3 planeNormal, in cone_t cone)
+{
+	const vec3 m = cross(cross(planeNormal, cone.direction), cone.direction);
+	const vec3 farthestBasePoint = cone.tip + (cone.direction * cone.height) - (m * cone.baseRadius); // farthest to plane's surface, away from positive half-space
+	return farthestBasePoint;
+}
+
+bool cullCone(in cone_t cone, in nbl_glsl_shapes_AABB_t aabb)
+{
+	float maxCosine = projectedSphericalVertex(cone.tip, cone.direction, vec3(aabb.minVx.x, aabb.minVx.y, aabb.minVx.z));
+
+	for (uint i = 1u; i < 8u; ++i)
+	{
+		const uvec3 t = (uvec3(i) >> uvec3(0, 1, 2)) & 0x1u;
+		const vec3 vertex = mix(aabb.minVx, aabb.maxVx, t);
+
+		// assuming cone.direction is normalized
+		maxCosine = max(projectedSphericalVertex(cone.tip, cone.direction, vertex), maxCosine);
+	}
+
+	const bool allVerticesOutsideCone = maxCosine < cone.cosHalfAngle;
+
+	if (cone.cosHalfAngle <= 0.f) // obtuse
+	{
+		return allVerticesOutsideCone; // cull if whole AABB is inside complementary acute cone
+	}
+	else if (any(or(greaterThan(aabb.minVx, cone.tip), greaterThan(cone.tip, aabb.maxVx))) && allVerticesOutsideCone)
+	{
+		// step 1
+		for (uint i = 0u; i < 8u; ++i)
+		{
+			const uvec3 t = (uvec3(i) >> uvec3(0, 1, 2)) & 0x1u;
+			const vec3 vertex = mix(aabb.minVx, aabb.maxVx, t);
+
+			const vec3 waypoint = normalize(vertex - cone.tip);
+			
+			// Todo(achal): nbl_glsl_slerp_impl_impl wants the length of cross of first two params to be exactly 1, hence there should be division
+			// by sine of angle between the cone.direction and waypoint, however, it introduces a chance of divison by zero as well which messes up
+			// things massively (as I have come to find out with a previous bug) --need to think about what to do when the sine approaches 0.
+			const vec3 normal = nbl_glsl_slerp_impl_impl(cone.direction, normalize(waypoint), sqrt(1.f - cone.cosHalfAngle * cone.cosHalfAngle));
+			if (dot(nbl_glsl_shapes_AABB_getFarthestPointInFront(aabb, normal) - cone.tip, normal) < 0.f)
+				return true;
+		}
+
+		vec4 planes[PLANE_COUNT] = getAABBPlanes(aabb);
+
+		for (uint i = 0u; i < PLANE_COUNT; ++i)
+		{
+			const vec3 normal = planes[i].xyz;
+
+			float farthestPoint = dot(normal, cone.tip);
+			if (dot(normal, cone.direction) < cone.cosHalfAngle)
+				farthestPoint = max(dot(normal, findFarthestPointOnConeInDirection(normal, cone)), farthestPoint); // https://www.3dgep.com/forward-plus/#Frustum-Cone_Culling 
+			else
+				farthestPoint += (cone.height / cone.cosHalfAngle);
+
+			if (farthestPoint < planes[i].w)
+				return true;
+		}
```

Comment:
AABB stands for **axis aligned** bounding box, only one component out of XYZ is ever non-zero, futhermore its only either +1 or -1, so no need for full dot products